rfcs/0000-cfb-encoding/0000-cfb-encoding.md

```diff
@@ -0,0 +1,138 @@
+
+## Format components
+
+The buffer consists of a root table offset and the data portion.
+
+The root table offset is the start position of the root table in the buffer.
+The data portion is a series of components, including tables, unions, strings, and vectors.
+
+All components must align by padding minimal amount of zeros before to meet
+the alignment requirements.
+
+### Data order
+
+All the components in the data portion construct a tree, in which the root table is the root node. Only the tables and vectors have children nodes:
+
+- The table's children are all the reference fields, including tables, unions, strings, and vectors.
+- The vector's children, straightforwardly, are the items in it when it is a vector of tables, unions or strings.
+
+The tree lays down into an array using depth-first traverse.
+
+```text
+                     +---+
+                     | 1 |
+                     +-+-+
+                       |
+                +-------------+
+                |      |      |
+              +-+-+  +-+-+  +-+-+
+              | 2 |  | 7 |  | 8 |
+              +-+-+  +---+  +-+-+
+                |             |
+         +------+             +------+
+         |      |             |      |
+       +-+-+  +-+-+         +-+-+  +-+--+
+       | 3 |  | 6 |         | 9 |  | 12 |
+       +-+-+  +---+         +-+-+  +----+
+         |                    |
+  +------+                    +-------+
+  |      |                    |       |
++-+-+  +-+-+                +-+--+  +-+--+
+| 4 |  | 5 |                | 10 |  | 11 |
++---+  +---+                +----+  +----+
+```
+
+The order of the children of a table is the same as in the schema. The vector children have the same sequence in the vector.
+
+The data portion must only contain the components in the tree.
+
+See the example [data-order].
+
+[data-order]: examples/data-order.md
+
+### Vtable order and deduplication
+
+The table can only refer to vtable before it. If the vtable already
+exists in the buffer before the table, the table must refer to the existing
+vtable. Otherwise, the vtable must be created and append to the buffer immediately before the table itself.
+
+In the example [data-order], vtables of `monster` and `stat` are just
+before the tables. However, `loots[1]` and `loots[2]` reuses the
+vtable of `stat` because of deduplication.
+
+### Data alignment
+
+The table's starting position `t` must be aligned, such that `t` is multiple of 4, and `t + 4` is the multiple of the most significant alignment of the present field in the table. All the fields in the table must be aligned as well.
+
+The vtable's starting position `t` must be a multiple of 2.
+
+The string's starting position `t` must be a multiple of 4.
+
+The vector's starting position `t` must be a multiple of 4, and `t + 4` is the multiple of the item alignment. **Specially**, if a vector of `ubyte` is marked as a nested buffer, `t + 4` must be multiple of 8.
```

Comment:
```suggestion
The vector's starting position `t` must be a multiple of 4, and `t + 4` must be the multiple of the item alignment. **Specially**, if a vector of `ubyte` is marked as a nested buffer, `t + 4` must be multiple of 8.
```

Reply:
> if a vector of `ubyte` is marked as a nested buffer, `t + 4` must be multiple of 8.

What's the reason? Can you give an example?

Reply:
See here https://github.com/dvidelabs/flatcc/blob/master/doc/binary-format.md#nested-flatbuffers

I think the rule is wrong here, since the nested buffer may contain a field that has a larger alignment requirement. It should be a configurable FlatBuffer user attribute.

We prefer nested buffers for tx and block header, which allows us to partially read and hash the block of memory directly. Otherwise, because of vtable sharing, the memory is not guaranteed to be consecutive.